Graduate conditional additional fields and add E2E testing to it.

senadir opened the pull request:
### Submission Review Guidelines:
This PR graduates additional fields outside of experimental feature flag and adds E2E tests.

Closes https://github.com/woocommerce/woocommerce/issues/54481
Closes https://github.com/woocommerce/woocommerce/issues/54473
<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. E2E tests should pass.
2. Add the following snippet somewhere
```php
add_action(
	'woocommerce_init',
	function () {
		woocommerce_register_additional_checkout_field(
			array(
				'id'         => 'plugin-namespace/alt-email',
				'label'      => 'Alternative Email',
				'location'   => 'contact',
				'type'       => 'text',
				'required'   => true,
				'validation' => array(
					'type'         => 'string',
					'format'       => 'email',
					'not'          => array(
						'const' => array( '$data' => '3/customer/billing_address/email' ),
					),
					'errorMessage' => array(
						'format' => 'Please enter a valid email address.',
						'not'    => 'Please enter a different email address to your contact email.',
					),
				),
			)
		);

		woocommerce_register_additional_checkout_field(
			array(
				'id'       => 'plugin-namespace/job-function',
				'label'    => 'What is your main role at your company?',
				'location' => 'contact',
				'required' => true,
				'type'     => 'select',
				'hidden'   => array(
					'customer' => array(
						'properties' => array(
							'billing_address' => array(
								'properties' => array(
									'email' => array(
										'pattern' => '^[a-zA-Z0-9._%+-]+@(gmail|hotmail|icloud)\.com$',
									),
								),
							),
						),
					),
				),
				'options'  => array(
					array(
						'label' => 'Director',
						'value' => 'director',
					),
					array(
						'label' => 'Engineering',
						'value' => 'engineering',
					),
					array(
						'label' => 'Customer Support',
						'value' => 'customer-support',
					),
					array(
						'label' => 'Other',
						'value' => 'other',
					),
				),
			)
		);
	}
);
```
3. On an incognito window, go to checkout with a product.
4. You should see email, alternative email, and "role" select, and no error from server 
5. Input an invalid email `test.com` in the alternative email, and blur the field, notice that it shows an error 
6. In the email field, enter a gmail email. Add the same email in the alternative email field, make sure it doesn't crash
7. Change your email domain to something else (a8c.com), notice the error is gone from the alternative email field, and a new select field is visible.
8. Select a value for the new select field, refresh the page, make sure it persisted. Also make sure the email did not revert from its previous value.
9. Change the alternative email to the same one as the primary email, an error would be shown.
10. Change the select to some other value, make sure it persists, refresh the page, it should still have persisted. (no issue marked for this, treating it as enhancement).
11. Change your primary email back to hotmail or gmail but different from alternative, notice that the select will now be hidden, and no validation errors are present.
12. Place the order. In the order confirmation screen, make sure `What is your main role at your company?` is not present because it was hidden. 

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Introduce conditional visibility and validation to Checkout additional fields.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

senadir committed:
Add E2E tests for additional fields schema rules

senadir committed:
Graduate feature

nielslange commented on plugins/woocommerce/client/blocks/tests/e2e/plugins/additional-checkout-fields.php:
```diff
@@ -1,13 +1,12 @@
 <?php
+declare(strict_types=1);
```
I see that you added `declare(strict_types=1);`, which is great! However, I noticed that the variables and functions are not typed. What do you think about adding type hints to fully benefit from strict typing? 

For example, changing:

```php
private $additional_checkout_fields_option_name = 'woocommerce_additional_checkout_fields';
```

to:

```php
private string $additional_checkout_fields_option_name = 'woocommerce_additional_checkout_fields';
```

And changing:

```php
public function enable_custom_checkout_fields() {}
```

to:

```php
public function enable_custom_checkout_fields(): void {}
```

If we do this, we should ensure all functions have return types, not just these examples.

nielslange commented on plugins/woocommerce/client/blocks/tests/e2e/tests/checkout/checkout.page.ts:
```diff
@@ -16,6 +16,7 @@ import {
 export class CheckoutPage {
 	private BLOCK_NAME = 'woocommerce/checkout';
```
```suggestion
```

Unrelated to your PR, but I noticed that this class property is no longer used and could be removed.

nielslange commented on plugins/woocommerce/client/blocks/tests/e2e/tests/checkout/additional-fields.guest-shopper.block_theme.spec.ts:
```diff
@@ -291,5 +292,129 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
+
+			await checkoutPageObject.waitForCheckoutToFinishUpdating();
+
+			await checkoutPageObject.waitForCheckoutToFinishUpdating();
```
```suggestion
			await checkoutPageObject.waitForCheckoutToFinishUpdating();
```

It looks like this call was duplicated, possibly by AI. Removing the extra call should be fine.

nielslange reviewed (dismissed):
Thanks for working on this, @senadir. The E2E tests are passing, and I didn't spot anything critical in your changes. However, I noticed that you applied `declare(strict_types=1);` to one file, and all other modified PHP files also have this definition. That said, most of the variables and functions in these files are not typed. While this isn't strictly related to your PR, it might be worth considering adding type hints for better consistency and stricter type safety.

That said, I'm pre-approving this PR as it works as expected.

gigitux committed:
[Product Collection]: Refactor InheritQueryControl and FilterableControl to use 'default' and 'custom' values for state management (#56582)

* Refactor InheritQueryControl and FilterableControl to use 'default' and 'custom' values for state management. Update onChange handlers to set boolean attributes.

* Add changefile(s) from automation for the following project(s): woocommerce-blocks

---------

Co-authored-by: github-actions <github-actions@github.com>

senadir committed:
Address review comments

dmallory42 committed:
[Payments NOX] Migrate to using core feature flag instead of WCA one (#55692)

* WIP: Migrating the feature flag.

* Linting

* Update beta tester changelog.

* Update beta tester changelog.

* Switch base branch to release/9.8 and merge

* Hook in the experiment to adjust default feature enablement

* test: E2E tests fix

* test: E2E fixes

* test: E2E fixes

* More e2e attempts

* More e2e attempts

* More e2e attempts

* More e2e attempts

* Add more guardrails around suggestion details

* More e2e tests adjustments

* Finally fix e2e tests

* Improve gateway ordering filtering

* Fix e2e tests

* Fix e2e tests

* Lint fix

* Fix some backward compatibility with existing Settings APIs

* Another e2e test try

* Another e2e test try

* Add delete option to e2e tests

* Add delete option to e2e tests

* One more e2e try

* e2e fix

* Fix e2e test 'cuz I'm...

* Clean-up e2e tests

* Revert "Clean-up e2e tests"

This reverts commit 7c188d0d351092657346e533f8ce895ad50ed28e.

* Fix comment typo

* Lint fixes

* Allow the feature flag option to be updated

---------

Co-authored-by: Vlad Olaru <vlad@pixelgrade.com>
Co-authored-by: Vlad Olaru <vlad.olaru@automattic.com>

senadir committed:
fix linting

Commit:
Add/56340 support install options (#56373)

* Add a new filter to get an instance of PluginsInstallLogger

* Use woocommerce_rest_api_plugins_install_logger filter to set a new logger to process install_options

* Accept source parameter

* Fix bug with $source -- it was not available in the async

* Set core-profiler as source

* Lint fixes

* Add woocommerce_plugins_install_before and woocommerce_plugins_install_after hooks

* Correct hook name

* Take care of JSON decode error

* Add unit test

* Lint fixes

* Add changefile(s) from automation for the following project(s): @woocommerce/data, woocommerce, woocommerce/client/admin

* Move woocommerce_plugins_install_before and woocommerce_plugins_install_after to OnBoardingSetupWizard class. I am also suppressing lint errors from unchanged, but old codes

* Document `source

* Remove install_priority property -- we always process options before installing a plugin

* Option can only be added. It should not allow updating existing options

* Update tests

* Lint fixes

* Fix json error handling

* Update plugins/woocommerce/src/Admin/PluginsHelper.php

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

* Update plugins/woocommerce/src/Admin/PluginsHelper.php

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

* Make sure options is an object

* Make the logger optional -- uses wc_get_logger if not provided

* Add test case for disallowed option

* Remove unused import

* Make sure options exist before converting to object

* Fix add_option argument order

* Update plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/ProcessCoreProfilerPluginInstallOptions.php

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

* Allow autoload value `no`

* Allow true and false

* Normalzie yes and no to bool

* Fix autoload value

* Use yoda

---------

Co-authored-by: github-actions <github-actions@github.com>
Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

Konamiman committed:
Deprecate the woocommerce_register_feature_definitions action (#56475)

* Deprecate the register_feature_definitions hook.

Also move all the existing feature registrations that used the hook
(HPOS and COGS) to FeaturesController::get_feature_definitions.

* Refactor FeaturesController to avoid infinite loops in DI resolutions

* Fix unit tests

* Remove usage of the woocommerce_register_feature_definitions hook in PaymentsController

* Add changelog file

* Try fixing failing CI.

* Revert "Try fixing failing CI."

This reverts commit e951f35a67ba298673171871d8f385cc90345820.

* Explicitly call FeaturePlugin::init, needed for CLI

---------

Co-authored-by: Vladimir Reznichenko <kalessil@gmail.com>

senadir committed:
Clear up hidden fields from final order

senadir committed:
Skip pushing fields that have an error but push other valid fields

senadir committed:
fix rebase

senadir committed:
remove already defined functiin

senadir committed:
Fix bug in which function is called with no values

senadir committed:
Fix bug in which function is called with no values

senadir committed:
Merge branch 'trunk' into add/e2e-tests-additional-fields

mikejolley commented on plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php:
```diff
@@ -934,7 +930,7 @@ public function validate_field( $field, $field_value ) {
 				sprintf(
 					'Field validation for %s encountered an error. %s',
-					esc_html( $field_key ),
+					esc_html( $field['id'] ),
```
This isn't blocking or anything but I feel we should validate somewhere that the field is actually a field because its just an array without a proper type i.e each field should have an ID. Do we do that somewhere?

senadir replied:
I agree, but we're the ones passing fields around and we validate them at registration.

mikejolley commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -165,12 +164,10 @@ protected function get_block_type_script( $key = null ) {
-		if ( Features::is_enabled( 'experimental-blocks' ) ) {
-			$checkout_fields = Package::container()->get( CheckoutFields::class );
+		$checkout_fields = Package::container()->get( CheckoutFields::class );
 			// Load schema parser asynchronously if we need it.
```
Indentation wrong here

mikejolley commented on plugins/woocommerce/src/StoreApi/Utilities/CheckoutTrait.php:
```diff
@@ -220,28 +219,26 @@ private function get_request_payment_method_title( \WP_REST_Request $request ) {
+			if ( isset( $additional_fields[ $key ] ) ) {
+				$this->additional_fields_controller->persist_field_for_order( $key, $posted_additional_fields[ $key ], $this->order, 'other', false );
+			} elseif ( '' === $value ) {
+				// Fields that are hidden (not on $additional_fields) but got posted with an empty value should be removed from the order.
```
What was this change for? Does it need additional test instructions?

senadir replied:
This probably got in during rebase or I added it at some point, I need to check and possibly delete it.

senadir replied:
This was me fixing a similar bug that you also fixed, i reverted my changes

mikejolley reviewed (changes requested):
Minor changes needed, but I also noticed some changes in logic as well as the removal of the feature flag. With that in mind I think the PR description needs to explain this and include some additional steps for testers to ensure this ships hassle free.

senadir committed:
address feedback

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

mikejolley reviewed (approved):
Thanks for adding more instructions. Tested as it should but I did spot one issue which could be done in a follow up.

If you change the "role" but quickly refresh, the change is lost because it didn't have time to persist. We should block navigation while pushing.

Other than that, no issues. This should be a feature highlight when live.

🚢

mikejolley commented:
CI might need retriggering.
